File: labosi/lab-1/2007-08/by_Kreso/IRG/lab1.py
from OpenGL.GL import *
from OpenGL.GLU import *
from OpenGL.GLUT import *

tocke=[]
prozor=[]

def draw():
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
    glFlush()

def idle():
    global angle
    angle += 2.0
    glutPostRedisplay()

def kbd(k, x, y):
    if ord(k) == 27: # Escape
        sys.exit(0)
    if k == 'r':
	glColor3f(1,0,0)
    if k == 'g':
	glColor3f(0,1,0)
    if k == 'b':
	glColor3f(0,0,1)
    if k == 'k':
	glColor3f(0,0,0)
    glRecti(prozor[0]-15, prozor[1]-15, prozor[0], prozor[1])
    glFlush()



def myLine(x0,y0,x1,y1):
    glBegin(GL_LINES)
    glVertex2i(x0,y0+20)
    glVertex2i(x1,y1+20)
    glEnd()
  
    # Bresenham
    steep = abs(y1 - y0) > abs(x1 - x0)
    if steep:
	x0, y0=y0, x0
	x1, y1=y1,x1
    if x0 > x1:
	x0, x1=x1,x0
	y0, y1=y1,y0
    deltax = x1 - x0
    deltay = abs(y1 - y0)
    error = 0.
    deltaerr = float(deltay) / deltax
    y = y0
    if y0 < y1:
	ystep = 1.
    else:
	ystep = -1.
	
    glBegin(GL_POINTS)
    for x in range(x0,x1):
	if steep:
	    glVertex2f(y,x)
	else:
	    glVertex2f(x,y)
	error +=deltaerr
	if error >= 0.5:
	    y +=ystep
	    error-=1.

    glEnd()



def mouse(button,state,x,y):
    if state == GLUT_DOWN:
	print 'click (%i, %i)' % (x,y)
    global tocke
    if button == GLUT_LEFT_BUTTON and state == GLUT_DOWN:
	tocke.append((x,y))
	if len(tocke) == 2:
	    print 'crtam liniju'
	    myLine(tocke[0][0],tocke[0][1],tocke[1][0],tocke[1][1])
	    tocke=[]
	    glFlush()

    elif button == GLUT_RIGHT_BUTTON and state == GLUT_DOWN:
	reshape(prozor[0],prozor[1])
	print 'brisem canvas'


def reshape(width, height):
    global prozor
    prozor=[width,height]
    h = float(height) / float(width)
    glViewport(0, 0, width, height)
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    gluOrtho2D(0, width, height, 0)
    glClearColor( 1., 1., 1., 0.)
    glClear( GL_COLOR_BUFFER_BIT )
    glPointSize(1.0)
    glColor3f(0., 0., 0.)
    glFlush()


glutInit(sys.argv)
glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB)
glutInitWindowSize(500,500)
glutInitWindowPosition(100,100)

window = glutCreateWindow("Glut OpenGL Linija")
glutReshapeFunc(reshape)
glutDisplayFunc(draw)
glutMouseFunc(mouse)
glutKeyboardFunc(kbd)
print "Lijevom tipkom misa zadaj tocke - algoritam Bresenham-a"
print "Tipke r, g, b, k mijenjaju boju."

glutMainLoop()

File: labosi/lab-1/2011-12/by_unknown_/lab1.cpp
//*********************************************************************************
//	Iscrtavanje linije 0-45, usporedba s glLine naredbom.
//	Pocetna i krajnja koordinata se zadaju lijevom tiplom misa, desnom se brise.
//	r,g, b, k s tastature mijenja boju.
//
//	Zadatak: Treba preraditi void myLine(GLint xa, GLint ya, GLint xb, GLint yb)
//	tako da radi ispravno za sve kutove
//
//*********************************************************************************

#ifdef _WIN32
#include <windows.h>             //bit ce ukljuceno ako se koriste windows
#endif

//  #include <GL/Gl.h>
//  #include <GL/Glu.h>    nije potrebno ako se koristi glut
#include <GL/glut.h>
#include <stdio.h>
#include <math.h>


//*********************************************************************************
//	Pocetna tocka Lx[1], Ly[1] Krajnja tocke Lx[2], Ly[2] linije.
//	Ix - stanje (zadana 1 ili 2 tocka)
//*********************************************************************************

GLdouble Lx[2], Ly[2];
GLint Ix;

//*********************************************************************************
//	Pokazivac na glavni prozor i pocetna velicina.
//*********************************************************************************

GLuint window; 
GLuint sub_width = 256, sub_height = 256;

//*********************************************************************************
//	Function Prototypes.
//*********************************************************************************

void myDisplay		();
void myReshape		(int width, int height);
void myMouse		(int button, int state, int x, int y);
void myKeyboard		(unsigned char theKey, int mouseX, int mouseY);


//*********************************************************************************
//	Glavni program.
//*********************************************************************************

int main(int argc, char ** argv)
{
	glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
	glutInitWindowSize(sub_width,sub_height);
	glutInitWindowPosition(100,100);
	glutInit(&argc, argv);

	window = glutCreateWindow("Glut OpenGL Linija");
	glutReshapeFunc(myReshape);
	glutDisplayFunc(myDisplay);
	glutMouseFunc(myMouse);
	glutKeyboardFunc(myKeyboard);
	printf("Lijevom tipkom misa zadaj tocke - algoritam Bresenham-a\n");
	printf("Tipke r, g, b, k mijenjaju boju.\n");

	glutMainLoop();
    	return 0;
}


//*********************************************************************************
//	Osvjezavanje prikaza. (nakon preklapanja prozora) 
//*********************************************************************************

void myDisplay()
{
 	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	glFlush();
}

//*********************************************************************************
//	Promjena velicine prozora.
//*********************************************************************************

void myReshape(int width, int height)
{
	sub_width = width;                      	//promjena sirine prozora
    sub_height = height;						//promjena visine prozora

	Ix = 0;										//	indeks tocke 0-prva 1-druga tocka
    glViewport(0, 0, sub_width, sub_height);	//  otvor u prozoru
	glMatrixMode(GL_PROJECTION);				//	matrica projekcije
	glLoadIdentity();							//	jedinicna matrica
	gluOrtho2D(0, sub_width, 0, sub_height); 	//	okomita projekcija
	glClearColor( 1.0f, 1.0f, 1.0f, 0.0f );		//	boja pozadine
	glClear( GL_COLOR_BUFFER_BIT );				//	brisanje pozadine
	glPointSize(1.0);							//	postavi velicinu tocke
	glColor3f(0.0f, 0.0f, 0.0f);				//	postavi boju
}

//*********************************************************************************
//	Crtaj moju liniju.
//*********************************************************************************

//ispravljen primjer koji radi za sve kuteve
void myLine(GLint xa, GLint ya, GLint xb, GLint yb)
{
	GLuint tempX, tempY;

	if (xa > xb) 
	{
		//prva tocka je uvijek ona lijevo
		tempX = xa;
		tempY = ya;
		xa = xb;
		ya = yb;
		xb = tempX;
		yb = tempY;
	}

	glBegin(GL_LINES);
	{
		glVertex2i(xa,ya+5);					//	crtanje gotove linije
		glVertex2i(xb,yb+5);
	}
	glEnd();
	
	GLint	x = xa;                        		
	GLint	y = ya;
	
	GLint	dx = xb - xa;
	GLint	dy = yb - ya;

	GLboolean yRaste = true;

	if (dy < 0)
	{
		//gledamo da li pravac raste ili pada po y
		//ako pada, uzimamo apsolutnu vrijednost
		dy = abs(dy);
		yRaste = false;
	}

	GLfloat d;

	//gledamo da li je tan(kutRast) < 1
	if (dy/(double)dx < 1) d=abs(dy/(double)dx)-0.5;
	else d = abs(dx/(double)dy) - 0.5;

	printf ("dx = %d; dy = %d; d = %f\n", dx, dy, d);

	glBegin(GL_POINTS);
	if (dy/(double)dx < 1)
	{  
		for (int i = 0; i <= dx; i++)	     	
		{
			glVertex2i(x,y);
			//printf ("x = %d; y = %d; d = %f\n", x, y, d);
			if (d > 0)					
			{
				if (yRaste)y++;	
				else y--;
				d=d-1.0;				
			}
			x++;
			d+=dy/(double)dx;
		}
	}
	else
	{
		for (int i = 0; i <= dy; i++)	     	
		{
			glVertex2i(x,y);
			//printf ("x = %d; y = %d; d = %f\n", x, y, d);
			if (d > 0)					
			{
				x++;					
				d=d-1.0;				
			}
			if(yRaste)y++;
			else y--;
			d+=dx/(double)dy;
		}
	}
	glEnd();
}

//primjer sa stranica lab. vjezbi koji radi za do 45 stupnjeva
void myOriginalLine(GLint xa, GLint ya, GLint xb, GLint yb)
{
	glBegin(GL_LINES);
	{
		glVertex2i(xa,ya+5);			//	crtanje gotove linije
		glVertex2i(xb,yb+5);				
	}
	glEnd();
	
	GLint	x = xa;                        		//	Bresenhamov algoritam do 45  
	GLint	y = ya;
	
	GLint	dx = xb - xa;
	GLint	dy = yb - ya;

	GLfloat	d=dy/(double)dx-0.5;

	glBegin(GL_POINTS);
	  for (int i = 0; i <= dx; i++)	     	
		{
			glVertex2i(x,y);			
			if (d > 0)					
			{
				y++;					
				d=d-1.0;				
			}
			x++;
			d+=dy/(double)dx;
		}
	glEnd();
}

//*********************************************************************************
//	Mis.
//*********************************************************************************

void myMouse(int button, int state, int x, int y)
{
	//	Lijeva tipka - crta pocetnu tocku ili liniju.
	if (button == GLUT_LEFT_BUTTON && state == GLUT_DOWN)

	{
		//	Pamti krajnju tocke linije.
		Lx[Ix] = x;							//	upisi tocku
		Ly[Ix] = sub_height - y;
		Ix = Ix ^ 1;						//	flip - druga tocka

		//	Crta prvu tocku ili liniju do druge tocke.
		if (Ix == 0)	myLine(Lx[0], Ly[0], Lx[1], Ly[1]);
		//if (Ix == 0)	myOriginalLine(Lx[0], Ly[0], Lx[1], Ly[1]);
		else			glVertex2i(x, sub_height-y);

		printf("Koordinate tocke %d: %d %d \n", Ix^1, x, y);

		glFlush();
	}

	//	Desna tipka - brise canvas. 
	else if (button == GLUT_RIGHT_BUTTON && state == GLUT_DOWN)
	{
		myReshape(sub_width, sub_height);
	}
}


//*********************************************************************************
//	Tastatura tipke - r, g, b, k - mijenjaju boju.
//*********************************************************************************

void myKeyboard(unsigned char theKey, int mouseX, int mouseY)
{
	switch (theKey)
	{
		case 'r':
			glColor3f(1,0,0);
			break;

		case 'g':
			glColor3f(0,1,0);
			break;

		case 'b':
			glColor3f(0,0,1);
			break;

		case 'k':
			glColor3f(0,0,0);
			
	}
	glRecti(sub_width-15, sub_height-15, sub_width, sub_height); // crta mali kvadrat u boji
	glFlush();
}

